DirectProgramming/DPC++FPGA/Tutorials/Features/fpga_reg/src/fpga_reg.cpp

```diff
@@ -0,0 +1,216 @@
+//==============================================================
+// Copyright Intel Corporation
+//
+// SPDX-License-Identifier: MIT
+// =============================================================
+#include <CL/sycl.hpp>
+#include <CL/sycl/intel/fpga_extensions.hpp>
+#include <iomanip>
+#include <iostream>
+#include <vector>
+#include "dpc_common.hpp"
+
+using namespace sycl;
+using namespace std;
+
+// Artificial coefficient and offset data for our math function
+constexpr size_t kSize = 64;
+constexpr std::array<int, kSize> kCoeff = {
+            1,  2,  3,  4,  5,  6,  7,  8,  9,  10, 11, 12, 13, 14, 15, 16,
+            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32,
+            33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48,
+            49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64};
+constexpr std::array<int, kSize> kOffset = {
+            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32,
+            49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64,
+            33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48,
+            16, 15, 14, 13, 12, 11, 10,  9,  8,  7,  6,  5,  4,  3,  2,  1};
+
+// The function our kernel will compute
+// The "golden result" will be computed on the host to check the kernel result.
+vector<int> GoldenResult(vector<int> vec) {
+  
+  // The coefficients will be modified with each iteration of the outer loop.
+  std::array coeff = kCoeff;
+  
+  for (int &val : vec) {
+    // Do some arithmetic
+    int acc = 0;
+    for (size_t i = 0; i < kSize; i++) {
+      acc += coeff[i] * (val + kOffset[i]);
+    }  
+
+    // Update coeff by rotating the values of the array
+    int tmp = coeff[0];
+    for (size_t i = 0; i < kSize - 1; i++) {
+      coeff[i] = coeff[i + 1];
+    }
+    coeff[kSize - 1] = tmp;  
+
+    // Result
+    val = acc;    
+  }
+  
+  return vec;
+}
+
+// Forward declaration of the kernel name
+// (This will become unnecessary in a future compiler version.)
```

Comment:
This is already the case.

Reply:
Not entirely true. While this will compile successfully without the forward declaration, the kernel name will appear badly mangled in the FPGA reports if the forward declaration is omitted. Forward declaration is recommended until the mangling issue is resolved.

Reply:
That is a good point.

Reply:
I should nonetheless update the comment to clarify the reason.